### PR TITLE
Support schema evolution across versions

### PR DESCRIPTION
**Pull request by naegelejd**
Introduces support for schema evolution across versions, including automated schema change detection and backward-compatible serialization (currently only C++ binary).

### CODE REVIEW

File: models/evolution/model_v0/_package.yml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+namespace: EvoTest
+
+cpp:
+  sourcesOutputDir: ../../../cpp/evolution/v0/lib
```
Comment (johnstairs) [inline models/evolution/model_v0/_package.yml:4]: In our other models, we have generated to a directory named `generated`. I think  we should be consistent.

File: cpp/test/generated/binary/protocols.cc [hunk 2]
```diff
@@ -706,7 +706,7 @@ template<typename T1, yardl::binary::Reader<T1> ReadT1, typename T2, yardl::bina
 namespace basic_types::binary {
 namespace {
 template<typename K, yardl::binary::Writer<K> WriteK, typename V, yardl::binary::Writer<V> WriteV>
-[[maybe_unused]] void WriteAliasedMap(yardl::binary::CodedOutputStream& stream, basic_types::AliasedMap<K, V> const& value) {
+[[maybe_unused]] static void WriteAliasedMap(yardl::binary::CodedOutputStream& stream, basic_types::AliasedMap<K, V> const& value) {
```
Comment (johnstairs) [inline cpp/test/generated/binary/protocols.cc:709]: Is the addition of `static` to make the internal linkage explicit (even though the function is in an unnamed namespace) or is there something else?

File: cpp/evolution/v2/lib/binary/protocols.cc [hunk 3]
```diff
@@ -0,0 +1,3957 @@
...
+    break;
+  }
+}
+
+void ProtocolWithChangesReader::ReadFloatToStringImpl(float& value) {
+  switch (version_) {
+  case Version::v1: {
+    std::string float_to_string;
+    yardl::binary::ReadString(stream_, float_to_string);
+    value = std::stof(float_to_string);
```
Comment (johnstairs) [inline cpp/evolution/v2/lib/binary/protocols.cc:2747]: Similarly, this doesn't give an informative error message if `float_to_string` is something like "Hello". It's hard for the user to know what the problem is, or even whether there is a bug in the tool.

File: tooling/internal/cpp/binary/binary.go [hunk 4]
```diff
@@ -84,18 +84,18 @@ func writeHeaderFile(env *dsl.Environment, options packaging.CppCodegenOptions)
 			fmt.Fprintf(w, "class %s : public %s, yardl::binary::BinaryWriter {\n", writerClassName, common.QualifiedAbstractWriterName(protocol))
 			w.Indented(func() {
 				w.WriteStringln("public:")
-				fmt.Fprintf(w, "%s(std::ostream& stream)\n", writerClassName)
+				fmt.Fprintf(w, "%s(std::ostream& stream, Version version = Version::Latest)\n", writerClassName)
```
Comment (johnstairs) [inline tooling/internal/cpp/binary/binary.go:87]: The name `Latest` could be misleading. Imagine you have a an existing codebase written against v1. Then v2 comes along and you just want to reference it to genereate the ability to write to it. In that case v1 is more "current" than "latest"

**Comment by johnstairs** [comment]
If you run `yardl generate --watch` and change the current model file, any warnings produced briefly appear and then are cleared.

**Comment by johnstairs** [comment]
If I run `yardl generate --watch` and change a previous model file, the change is ignored.

File: tooling/pkg/dsl/evolution.go [hunk 5]
```diff
@@ -0,0 +1,1505 @@
...
+		args[i] = fmt.Sprintf("%s: %s", param.Name, arg)
+	}
+	argString := ""
+	if len(args) > 0 {
+		argString = fmt.Sprintf("<%s>", strings.Join(args, ", "))
+	}
+	return fmt.Sprintf("%s%s", baseName, argString)
+}
+
+// Returns the base definition for a NamedType
```
Comment (johnstairs) [inline tooling/pkg/dsl/evolution.go:341]: Might want to clarify unless the base is a primitive or generic type parameter

File: tooling/pkg/dsl/evolution.go [hunk 6]
```diff
@@ -0,0 +1,1505 @@
...
+		return out
+	}
+
+	newDef := resolveGenericDefinitions(newTd, oldTd)
+	oldDef := resolveGenericDefinitions(oldTd, newTd)
+	return &DefinitionPair{oldDef, newDef}
+}
+
+// Determine semantic equivalence between Old and New TypeDefinitions
+func getCrossModelSemanticEquivalence(newTds, oldTds []TypeDefinition) map[string]map[string]bool {
```
Comment (johnstairs) [inline tooling/pkg/dsl/evolution.go:400]: Not used?

**Comment by johnstairs** [comment]
I think we should warn if a protocol is renamed/removed

**Comment by johnstairs** [comment]
If I remove a field from a record, I get a compilation error saying that the variable that we create for writing the value is uninitialized. Also, the message "Removing non-Optional field 'xyz' may result in undefined behavior with previous versions" is not so great. I think we should have defined behavior in all cases.

File: tooling/internal/cpp/binary/binary.go [hunk 7]
```diff
@@ -451,15 +505,258 @@ func writeRwFunctionTemplateDeclaration(t dsl.TypeDefinition, w *formatting.Inde
...
+
+		fmt.Fprintf(w, "switch (%s.index()) {\n", sourceName)
+		w.Indented(func() {
+			for i := range tc.OldType().(*dsl.GeneralizedType).Cases {
+				fmt.Fprintf(w, "case %d: {\n", i)
+				w.Indented(func() {
+					if tc.OldMatches[i] {
+						fmt.Fprintf(w, "%s = std::get<%d>(%s);\n", targetName, i, sourceName)
+					} else {
+						fmt.Fprintf(w, "throw new std::runtime_error(\"Union type incompatible with previous version of model\");\n")
```
Comment (johnstairs) [inline tooling/internal/cpp/binary/binary.go:613]: is it possible to provide a path or desciption of the union?

**Comment by johnstairs** [comment]
Changing a type from 
```
[int, float]
```
to 
```
[int, float]
```

Seems to be allowed but the serialization to the previous version is broken.

Changing from `int` to `int*`, however,  is validation error.

Changing from 
```
[int, bool]
```
to 
```
[int, float]
```
 Results in C++ compilation errors.

**Comment by johnstairs** [comment]
When changing from int32 to int64 and encountering an overflow, I think we should raise and error.

**Comment by johnstairs** [comment]
Seems like this could be considered compatible:
image

**Comment by johnstairs** [comment]
Also this could be accepted (with warning of possible overflow):
image

**Comment by johnstairs** [comment]
Same here:
image

**Comment by johnstairs** [comment]
If you change the numerical values of an enum, we don't seem to do anything about it in the serializers.

**Comment by johnstairs** [comment]
image


Results in compilation error:
```
protocols.cc:70:11: error: 'a' may be used uninitialized [-Werror=maybe-uninitialized]
   70 |     value = std::move(a);
      |     ~~~~~~^~~~~~~~~~~~~~
```

**Comment by johnstairs** [comment]
I think this should be accepted:
image

**Comment by johnstairs** [comment]
Swapping the union cases and writing to the previous format seems to write the same binary data:
image

**Comment by johnstairs** [comment]
I'm wondering why it's ok to add a stream to a protocol but not to remove one.

**Comment by johnstairs** [comment]
I think it's getting confused with this:
image
When writing in the old format, it writes out the same as in the new format, effectively swapping A and B types

File: cpp/test/generated/binary/protocols.cc (hunk 2 above)
Comment (naegelejd) [reply cpp/test/generated/binary/protocols.cc:709]: My misconception about compiler optimizations for static vs unnamed namespace. This change has been reverted.

**Comment by naegelejd** [comment]
I have addressed all of the feedback above except for the following items, which we plan to support in the future:


| Missing support for: | Effort | Explanation |
|-|-|-|
| Scalar <-> Vector/Array conversions | Low | This is a reasonable schema change yardl should support |
| Removing Protocol steps | Medium | This is a reasonable schema change yardl should support |
| Changing generic Record type arguments | High | Requires making type conversion dynamic (e.g. template functions) so that a generic Record serializer can convert fields based on type arguments at runtime |
| Testing runtime conversion errors | Medium | Requires 2+ schema versions, codegen, compilation, and GTest for each scenario under test |

I plan to open an issue for each of these and address them after I finish implementing support for Matlab.

**Comment by johnstairs** [comment]
I think we should allow `complexfloat32 <-> complexfloat64`

File: docs/cpp/evolution.md [hunk 8]
```diff
@@ -0,0 +1,244 @@
+# Schema Evolution
+
+Data models often change over time, thus, yardl provides support for evolving your schema in a cross-version compatible manner.
+
+When deploying protocol readers and writers generated from the current version of your schema, you may need the ability to read and write older versions of your data.
+Conversely, you may want to update older versions of your software to read and write the current version.
+
+Yardl supports cross-version compatibility by *statically* generating conversions to and from other versions of your schema (i.e. at yardl codegen time).
+
+In the future, yardl will support *dynamic* compatibility between schema versions, enabling backward- and forward-compatibility between any schema version at runtime.
```
Comment (johnstairs) [inline docs/cpp/evolution.md:10]: We should also call out that for the moment, we only support this for the binary encoding format using C++.

**Comment by johnstairs** [comment]
If I change from a `float32` to an `int64`, the runtime checks are not what I would expect. When writing to the previous version, we check if the integer value falls within the range of a float32, which it always will. Then when reading the older version, we simply do a cast, when that is what might overflow.

**Comment by naegelejd** [comment]
> If I change from a `float32` to an `int64`, the runtime checks are not what I would expect. When writing to the previous version, we check if the integer value falls within the range of a float32, which it always will. Then when reading the older version, we simply do a cast, when that is what might overflow.

I will get this right. Too bad this isn't in the standard library: https://www.boost.org/doc/libs/1_84_0/libs/numeric/conversion/doc/html/boost_numericconversion/improved_numeric_cast__.html

**Review by johnstairs — APPROVED** [review]
Ship it!

**Pull request merged**